cotix/_collision_resolution.py

```diff
@@ -13,68 +13,253 @@
     perpendicular = perpendicular_vector(unit_collision_vector)
     change_of_basis = jnp.array([unit_collision_vector, perpendicular])
     change_of_basis_inv = jnp.linalg.inv(change_of_basis)
 
-    v1q = jnp.dot(body1.velocity, unit_collision_vector)
-    v1r = jnp.dot(body1.velocity, perpendicular)  # stays constant
-    v2q = jnp.dot(body2.velocity, unit_collision_vector)
-    v2r = jnp.dot(body2.velocity, perpendicular)  # stays constant
+    # everything below should be in the new coordinate system
+    v1_col_basis = change_of_basis @ body1.velocity
+    v2_col_basis = change_of_basis @ body2.velocity
 
-    v1q_new, v2q_new = _1d_elastic_collision_velocities(
-        body1.mass, body2.mass, v1q, v2q
+    v_rel = v1_col_basis - v2_col_basis
+
+    (
+        lever_arm1,
+        lever_arm2,
+        tangential_relative_velocity,
+    ) = _compute_lever_arms_and_tangential_velocity(
+        change_of_basis, contact_point, body1, body2
     )
 
-    v1qr_new = jnp.array([v1q_new * elasticity, v1r])
-    v2qr_new = jnp.array([v2q_new * elasticity, v2r])
+    # check that the collision has to be resolved
+    point1_v = body1.velocity + body1.angular_velocity * lever_arm1
+    point2_v = body2.velocity + body2.angular_velocity * lever_arm2
+    # todo: this might not work with super concave bodies,
+    #  when the center of mass is on the wrong side of the collision point
+    velocities_away = jnp.dot(point1_v - point2_v, body1.position - body2.position) >= 0
 
-    v1_new = jnp.matmul(change_of_basis_inv, v1qr_new)
-    v2_new = jnp.matmul(change_of_basis_inv, v2qr_new)
+    # called iff velocities_away is False
+    def continue_resolution(body1, body2):
+        # impulse computation is in accordance with
+        # https://github.com/knyazer/RSS/blob/
+        # 1246e03c5950a5549a128fbce97c7bd402f9bed7/engine/source/env/World.cpp#L87
 
-    body1 = body1.set_velocity(v1_new)
-    body2 = body2.set_velocity(v2_new)
+        # inertia is kg * m^2, so this is kg^-1
+        impulseFactor1 = (1 / body1.mass) + (lever_arm1**2) / body1.inertia
+        impulseFactor2 = (1 / body2.mass) + (lever_arm2**2) / body2.inertia
 
-    body1, body2 = _split_bodies(body1, body2, epa_vector)
-    return body1, body2
+        # here, we only care about the collision along the collision normal
+        # units are kg * m / s
+        primary_col_impulse = (
+            (1 + elasticity) * v_rel[0] / (impulseFactor1 + impulseFactor2)
+        )
+
+        # v1 dot perpendicular tries to be the same as v2 dot perpendicular.
+        #   clipped to be in the range [-friction_impulse_max, friction_impulse_max]
+        friction_impulse_max = friction_coefficient * jnp.abs(primary_col_impulse)
+        # this adds an impulse perpendicular to the collision normal
+        friction_impulse = jnp.clip(
+            # I am not too sure that the formula is correct here
+            tangential_relative_velocity / (impulseFactor1 + impulseFactor2),
```

Comment:
please take a look here. I am fairly sure about the signs, but the magnitude may be off here

Reply:
cause basically all my tests only look at the signs: > or < 0, > or < than before, etc.

testing for energy conservation is not fair I think, as in not perfectly elastic collisions kinetic energy is not conserved

Reply:
oki, will take a look in a few hours

Reply:
well in terms of units it seems like you have m*s^-1 * kg which is the unit for impulse.

Reply:
yeah, i have no idea how physics works. Theoretically should be alright, since you bounded the friction impulse by the value that makes sense (a multiple of newtonian friction force). The problem in our case I guess is that instantenous collisions don't really allow for friction in standard model: since friction is a force, the impulse would be force by time interval, and time interval is 0 for instantenous collisions. However, we would love to simulate something that would be 'similar' to the friction in the real world. For now I guess what you have done is a decent enough approximation

Reply:
To get a better approximation we probably need to have a look into brax/dojo/any other simulator with spring-y constraints. The first thing that comes to my mind without it, is additionally scaling by relative velocity: the higher relative velocity is, the smaller should be friction. But then, it actually should be bigger, because the force is bigger, so maybe it should be just a small constant-reduction-in-speed for every collision? idk. Anyway, good enough for now